### DISCUSSION TITLE
How do I apply a custom window to rolling function

### DISCUSSION THREAD

**Discussion by delahoussou** [pydata/xarray#7641] [Q&A]
I am working on this project. I have netcdf4 files. I calculated the amount of precipitation I have over every 6H time period. Out of that I looked at every instance where I have subsequent period of time where the precipitation is greater than 5 mm, I add those up. Then if the summation of those at a certain time is greater than 50mm ,I would like to see that. My goal is to see where we may have potential flooding event in the data. Everything works fine except that some dates are repeated more than twice which tell me that my code is not doing the summation of the events greater than 5mm. Let me explain what which some numbers. Let's say the amount of precipitation I have is this list [2,3,5,7,0,6,3,5,7,9,8,0]
Then my summation of event 5mm will be:
5 + 7 = 12
6
5 + 7 + 9 + 8 = 29
This is an idea of what I am seeking to do so that in my output I return the summations greater than 50 mm. I have attached pictures of my code and the output
![A6F043B2-7774-42F0-9926-2DFC1E924F88](https://user-images.githubusercontent.com/101612226/226075397-f5767709-11ed-4c2a-b08b-cf03b962f005.jpeg)
![D8FE3A30-7088-443D-B2A5-0E5BAC17122F](https://user-images.githubusercontent.com/101612226/226075399-7fa1e26e-77b1-4474-bd3a-bb9dce7fa4a1.jpeg)
![F7BB56A7-D85F-4929-A640-51C7DE78E192](https://user-images.githubusercontent.com/101612226/226075401-172fbf94-68d0-4ed4-a7c6-66f201a9c638.jpeg)

**Comment by keewis** [comment]
First of all, let's define some sample data that we can use to verify that the algorithm works without accessing your pile of netcdf files:
```python
In [2]: import xarray as xr
   ...: import numpy as np
   ...: 
   ...: # I'll use the name `precipitation` here, but you can substitute that for `msc_prep`
   ...: precipitation = xr.Dataset(
   ...:     {
   ...:         "station_a": ("time", [6, 7, 8, 2, 1, 9, 8, 1, 7]),
   ...:         "station_b": ("time", [2, 7, 5, 4, 9, 5, 3, 8, 1]),
   ...:         "station_c": ("time", [7, 9, 1, 4, 6, 3, 8, 6, 7]),
   ...:         "time": xr.date_range("2001-01-01", periods=9, freq="6H"),
   ...:     }
   ...: ).to_array(dim="station", name="precipitation")
   ...: precipitation
Out[2]: 
<xarray.DataArray 'precipitation' (station: 3, time: 9)>
array([[6, 7, 8, 2, 1, 9, 8, 1, 7],
       [2, 7, 5, 4, 9, 5, 3, 8, 1],
       [7, 9, 1, 4, 6, 3, 8, 6, 7]])
Coordinates:
  * time     (time) datetime64[ns] 2001-01-01 2001-01-01T06:00:00 ... 2001-01-03
  * station  (station) object 'station_a' 'station_b' 'station_c'
```

As an aside, while you're iterating on the issue (or even later) I'd suggest extracting the data from your huge netcdf files into a single, smaller one (that way, you don't have to wait while the massive, unchunked files are being loaded and subset):
```python
# run once:
msc_prep = prsent_prep["PREC_ACC_NC"].isel(
    south_north=xr.DataArray(station_indices[:, 0], dim="station"),
    west_east=xr.DataArray(station_indices[:, 1], dim="station"),
).sel(Time=slice("2000", "2013"))
msc_prep.to_netcdf("msc_prep.nc")

# then load from the small file
msc_prep = xr.open_dataset("msc_prep.nc", engine="netcdf4").compute()
```

Since this is a simpler version than the code we discussed during the office hours on Friday, I actually think we don't need the rolling operation. Instead, we can use a cumulative sum to construct labels for each consecutive rainfall event:
```python
In [6]: rainfall_events = precipitation >= 5
   ...: labels = (
   ...:     np.logical_not(rainfall_events)
   ...:     .cumsum(dim="time")
   ...:     .where(rainfall_events)
   ...:     .rename("rainfall_events")
   ...: )
   ...: labels
Out[6]: 
<xarray.DataArray 'rainfall_events' (station: 3, time: 9)>
array([[ 0.,  0.,  0., nan, nan,  2.,  2., nan,  3.],
       [nan,  1.,  1., nan,  2.,  2., nan,  3., nan],
       [ 0.,  0., nan, nan,  2., nan,  3.,  3.,  3.]])
Coordinates:
  * time     (time) datetime64[ns] 2001-01-01 2001-01-01T06:00:00 ... 2001-01-03
  * station  (station) object 'station_a' 'station_b' 'station_c'
```
We could create unique labels by prepending the station name, but that would mean we might have to do complex string operations to recreate the indexes. Instead, we'll use `flox`'s multi-dimensional `groupby` operations:
```python
In [7]: import flox.xarray
   ...: 
   ...: flox.xarray.xarray_reduce(precipitation.astype(float), "station", labels, func="sum")
Out[7]: 
<xarray.DataArray 'precipitation' (station: 3, rainfall_events: 4)>
array([[21., nan, 17.,  7.],
       [nan, 12., 14.,  8.],
       [16., nan,  6., 21.]])
Coordinates:
  * station          (station) object 'station_a' 'station_b' 'station_c'
  * rainfall_events  (rainfall_events) float64 0.0 1.0 2.0 3.0
```
In order to also get the time span for each group, I *think* we have to do multiple `groupby`s on the time coordinate (not sure, though, there might be a better way):
```python
In [18]: _, time = xr.broadcast(precipitation, precipitation.time)
    ...: 
    ...: start = flox.xarray.xarray_reduce(time, "station", labels, func="min")
    ...: end = flox.xarray.xarray_reduce(time, "station", labels, func="max")
    ...: xr.Dataset({"start": start, "end": end})
Out[18]: 
<xarray.Dataset>
Dimensions:          (station: 3, rainfall_events: 4)
Coordinates:
  * station          (station) object 'station_a' 'station_b' 'station_c'
  * rainfall_events  (rainfall_events) float64 0.0 1.0 2.0 3.0
Data variables:
    start            (station, rainfall_events) datetime64[ns] 2001-01-01 ......
    end              (station, rainfall_events) datetime64[ns] 2001-01-01T12:...
```
(@dcherian, any idea on how to simplify that, and how to avoid the potentially very sparse result? Maybe constructing unique rainfall event labels, then creating a multi-index from that is the way?)

cc @scottyhq

---

For the more complex condition you had which also took the succeeding value into account we'd use `rolling` to construct `rainfall_events`, but I guess that's a follow-up question.

**Reply by delahoussou** [reply]
Thank you for the answer. I did try. 
First I realized in line 6 you called out "labels" to see what was inside it returned the output. 
I did the same thing in my code. It run well but it did not generate any output. 
Also in my environment I could not run the import statement you have in line 7. 
I retained my for loop at the beginning since I am iterating over 4 different locations. 
Can you double check the code for me. Thank you. 

![4341E203-764A-4C81-B0D6-644D1AE8A6D0](https://user-images.githubusercontent.com/101612226/226497039-ec73aebe-a9b6-4615-962a-45c47900afab.jpeg)

**Reply by keewis** [reply]
to clarify, the code that extracts the stations and writes to disk using `to_netcdf` should be put into its own cell, run once (so the file exists) and then commented out / changed to raw (change the drop-down from "Code" to Raw", or type <kbd>Esc</kbd>, <kbd>r</kbd> – i.e. type <kbd>r</kbd> in command mode). The idea is to cache the step in your calculation that takes the longest (extracting the data of interest), so all subsequent runs can make use of a small file that fits entirely in memory.

The code is also intentionally written in a way that lets us get rid of the manual loop: `msc_prep` contains all the stations / pixels of interest, and the `groupby` accumulates the results per station.

So really, what I was thinking is this (I'm marking omitted code with `...` so I don't have to type as much, just fill in the code you already have):
```python
# cell 1: run once to create "msc_prep.nc" (or whatever name you prefer), then deactivate
muscatine = ...
...
station_indices = np.array((muscatine, ...))

prsent_prep = xr.open_mfdataset(...)
msc_prep = prsent_prep["PREC_ACC_NC"].isel(
    south_north=xr.DataArray(station_indices[:, 0], dims="station"),
    west_east=xr.DataArray(station_indices[:, 1], dims="station"),
).sel(Time=slice("2000", "2013"))
msc_prep.to_netcdf("msc_prep.nc")

# cell 2: analysis
msc_prep = xr.open_dataset("msc_prep.nc", engine="netcdf4").compute()

rainfall_events = msc_prep >= 5
event_labels = np.logical_not(rainfall_events).cumsum(dim="Time").where(rainfall_events).rename("rainfall_events")

# install `flox` for this to work
# e.g. using `mamba install -c conda-forge flox` or `conda install -c conda-forge flox`
# or however you install packages into your environment
import flox.xarray

_, time = xr.broadcast(msc_prep, msc_prep.Time)

rain = flox.xarray.xarray_reduce(msc_prep.astype(float), labels, func="sum")
start = flox.xarray.xarray_reduce(time, labels, func="min")
end = flox.xarray.xarray_reduce(time, labels, func="max")

results = xr.Dataset({"rain": rain, "start": start, "end": end})
```

Also, feel free to add `display` or `print` calls to the code to see what's happening (e.g. `display(labels)`) – the difference is that with `display` you'll get the fancy interactive `xarray` reprs.

**Reply by keewis** [reply]
Most of the work done here is in two lines. With the first line:
```python
rainfall_events = msc_prep >= 5
```
we check every timestep for every station to see if there was a sufficient amount of rain. The result is a 2D boolean array with dimensions `Time` and `station`.

In the second line we first invert that, which gives us the elements that did *not* meet the criteria: those are the boundaries between each consecutive event. If we then compute the cumulative sum along `Time` we get a value per consecutive rainfall event per station (because remember: `rainfall_events` is a 2D array). However, that means that the boundary *before* the rainfall event has the same number as the rainfall event itself, so we mask out the boundaries using `where`.
```python
event_labels = (
    rainfall_events
    .pipe(np.logical_not)  # invert to get the boundaries
    .cumsum(dim="Time")  # cumulative sum to get a number for each rainfall event per station
    .where(rainfall_events)  # get rid of the boundaries
    .rename("rainfall_events")
)
```

(note that `rainfall_events.pipe(np.logical_not)` is a different way of writing `np.logical_not(rainfall_events)`, but does exactly the same thing)

**Reply by keewis** [reply]
Something that just occurred to me is that if we have multiple "boundary" elements in sequence we increase the label way too fast. To get rid of that, we can use `rolling`:
```python
def detect_boundary(arr, axis):
    # detect boundaries, which are `False` followed by `True`
    # The idea is that the sum has to be `1` along the window axis (i.e. one `True` and one `False`)
    # and the last value along the window axis has to be `True`
    return (np.sum(arr, axis=axis) == 1) & np.take(arr, axis=axis, indices=-1)

event_labels = (
    rainfall_events
    .pad({"Time": (1, 0)}, mode="reflect")  # pad so we don't get `nan` for the first value
    .rolling(Time=2)  # inspect adjacent values together
    .reduce(detect_boundary)  # return `True` only if we see a boundary followed by a event
    .isel(Time=slice(1, None))  # get rid of the padding
    .cumsum(dim="time")
    .where(rainfall_events)
    .rename("rainfall_events")
)
```

The first thing we do here is pad `rainfall_events` by prepending the first value per station, which allows us to keep the first value as it is. To make sure we only take one boundary between rainfall events into account (even if there are multiple consecutive "dry" timesteps), we then detect boundaries using `rolling` and a custom function. After that, we can get rid of the padding using `isel` and continue on as usual, with the difference that we don't have to call `np.logical_not` since the boundary detection already returns only the boundary elements.

---

My recommendation would be to make sure you first understand the algorithm itself before trying to understand this extension which aims to reduce the total number of labels.

**Reply by delahoussou** [reply]
Thank you for the answer. I am trying the code. There is one thing though. The first cell throws a MemoryError: cannot allocate 11GB to disk 
I read the xarray documentation but it does not talk about how to handle something like that. I however saw an example where in to_netcdf() function, they specified "engine" , it is also in the documentation. So I used something like that.  I have: 

msc_prep.to_netcdf("msc_prep.nc", engine = 'scipy')  

It did work but I wanna know if: 
1.)  that is the correct thing to do. And second after I did that .rename('rainfall_events")  to handle the MemoryError
2.) After I did that, the cell ran properly but then .rename()  also threw an error saying that a dictionary has to be specified to .rename() instead of the string that we currently passed to it.

**Reply by delahoussou** [reply]
![Screenshot (8)](https://user-images.githubusercontent.com/101612226/227811916-2b8a8923-f08c-40a6-9b17-d3d4808943b7.png)

**Reply by keewis** [reply]
apologies, there was a typo in the code I posted (it's corrected now).

If the first cell worked, that's great! `engine="scipy"` means we use a different library (and file format) to write the results, but that's perfectly fine, especially since the resulting file should be pretty small. However, I'm not sure I understand why it worked when `engine="netcdf4"` (the default, if you have `netcdf4` installed) does not. Can you post the code you used?

The `rename` only works like that if `rainfall_events` is a `DataArray`, so of course we have to change how we compute that (my mistake, I should have known). So either you change the call to `.rename_vars({"PREP_ACC_NC": "rainfall_events"})`, or we use:
```python
rainfall_events = msc_prep["PREP_ACC_NC"] >= 5
```
The first uses the `Dataset` syntax to rename variables, while the second option results in `rainfall_events` being a `DataArray` object so the original `rename` works.

**Reply by delahoussou** [reply]
```python
muscatine = np.array([559, 821])
algona = np.array([601, 754])
cedar_fall = np.array([588, 791])
Pella = np.array([557, 783])
station_indices = np.array((muscatine, algona, cedar_fall, Pella))
msc_prep = prsent_prep["PREC_ACC_NC"].isel(
        south_north=xr.DataArray(lat_lon[:, 0], dims = "station_indices"),
        west_east=xr.DataArray(lat_lon[:, 1], dims = "station_indices"), 
    ).sel(Time=slice("2000", "2013"))
msc_prep.to_netcdf("msc_prep.nc", engine = 'scipy')
```

**Reply by delahoussou** [reply]
Here is the code I used. Not passing  engine = "scipy" argument in the last line  will through a MemoryError.

**Reply by delahoussou** [reply]
Or if you don't pass any argument apart from the string "msc_prep"

**Reply by keewis** [reply]
okay, so that was still wrong, it should be `station_indices` instead of `lat_lon`. Could you post the `repr` of `msc_prep` before calling `.to_netcdf`? That might help with figuring out why it doesn't work.

I'd expect `rainfall_events` to be either a `DataArray` with dimensions `Time` × `station_indices`, or the equivalent as a `Dataset`.

**Reply by delahoussou** [reply]
In the first portion, I think you mean instead of

```python
msc_prep = prsent_prep["PREC_ACC_NC"].isel(
        south_north=xr.DataArray(lat_lon[:, 0], dims = "station_indices"),
        west_east=xr.DataArray(lat_lon[:, 1], dims = "station_indices"), 
    ).sel(Time=slice("2000", "2013"))
```
it should rather be: 
```python  
station_indices = np.array((muscatine, algona, cedar_fall, Pella))
msc_prep = prsent_prep["PREC_ACC_NC"].isel(
        south_north=xr.DataArray(station_indices[:, 0], dims = "station_indices"),
        west_east=xr.DataArray(station_indices[:, 1], dims = "station_indices"), 
    ).sel(Time=slice("2000", "2013"))
```

**Reply by delahoussou** [reply]
![FD2DAE0B-6DF0-4729-A6E4-473E31E1C8AC](https://user-images.githubusercontent.com/101612226/228000219-b05db073-fc27-4357-8bc8-aad01241c95d.jpeg)

**Reply by delahoussou** [reply]
This is the result I have before writing it to netcdf4

**Reply by delahoussou** [reply]
Just something to think about. In the image, the last line msc_prep, I am wondering if that is what is running in the code or if that what I wrote to disk. If you look at the left side, you will see a msc_prep.nc which we created early, so I am wondering it that is what is still showing in the last line or in the whole code is running the showing the result. 
I think what you are looking for is to see what format is msc_prep before we write it to netcdf4 with to_netcdf. 
Something to think about.

**Reply by keewis** [reply]
what I wanted to know was whether the output data was correct (which it appears to be). We could even call `.compute()` on the data before saving to disk because it is so small now. I still don't get why you can only save to netcdf3 (the `scipy` engine only writes that format), but in any case it appears to have worked as I intended it to: extract the timeseries for the four stations.

Yes, the file is only appended to, so if you want to try different versions you'd have to add `mode="w"` to the `.to_netcdf` call:
```python
msc_prep.to_netcdf("msc_prep.nc", mode="w", engine="scipy")
```

Once that works, did you try the second cell with the changes from the second paragraph of https://github.com/pydata/xarray/discussions/7641#discussioncomment-5438513?

**Reply by delahoussou** [reply]
It worked up to where we are calling the xarray_reduce. I don't understand. In the xarray.DataArray documentation, is said name was optional which is the reason why I believe you did not specify it: 
name ([str](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.str.html#xarray.DataArray.str) or [None](https://docs.python.org/3/library/constants.html#None), optional) – Name of this array.

But in xarray_reduce it says throw AttributeError is name is not specified. Is that a discrepancy between these two functions. You can look at @dcherian example in how he applied the xarray_reduce. I have also attached a picture of my error

**Reply by delahoussou** [reply]
![CF11A9CF-E977-4665-8132-7AF4CEE281C5](https://user-images.githubusercontent.com/101612226/228437740-4c5073ee-07f8-4d1c-b246-19e6d8bcbec6.jpeg)
![2F8F3E13-0602-4DBF-A467-9A6E87AD890E](https://user-images.githubusercontent.com/101612226/228437993-1124ccde-c6b9-4deb-8bf0-ead1ad3b4c9b.jpeg)

**Reply by keewis** [reply]
ah, yes. That tells you that it expected a `DataArray` instead of a `Dataset` (we should definitely make that a bit easier to understand, though). So I guess we should just change the call to `open_dataset` to:
```python
msc_prep = xr.open_dataset("msc_prep.nc", engine="netcdf4")["PREP_ACC_NC"].compute()
```
which should get rid of that issue (and you can use the `.rename` again to make the results a bit nicer).

**Reply by delahoussou** [reply]
Let me try that rn and get back to you.

**Reply by delahoussou** [reply]
xr.open_dataset is what we are already using. I guess you meant something else. Can you come again?

**Reply by delahoussou** [reply]
Do you mean open_dataarray? 

msc_prep = xr.open_dataarray("msc_prep.nc", engine="netcdf4")["PREP_ACC_NC"].compute() 
instead of 

msc_prep = xr.open_dataset("msc_prep.nc", engine="netcdf4")["PREP_ACC_NC"].compute()  ????

**Reply by keewis** [reply]
the difference is the `["PREP_ACC_NC"]` that I inserted. The line you had before is:
```python
msc_prep = xr.open_dataset("msc_prep.nc", engine="netcdf4").compute()
```
So we're pulling out a single variable from the dataset and discard any others (but as far as I remember that's the only variable, anyways)

**Reply by delahoussou** [reply]
ok

**Reply by keewis** [reply]
you can use `open_dataarray` instead as well, though

**Reply by delahoussou** [reply]
That worked

**Reply by delahoussou** [reply]
![110B0BF9-2F23-49F1-A21C-77D0F2DD5F13](https://user-images.githubusercontent.com/101612226/228471728-08eb3e8e-7011-4886-ade0-647fb1fe7a0a.jpeg)

**Reply by keewis** [reply]
that's because when you created `time` you're missing a few characters at the beginning of the line. It should be:
```python
_, time = xr.broadcast(msc_prep, msc_prep.Time)
```
(note the `_, `). This is important, because `broadcast` returns a tuple containing *all* the broadcasted arguments, and we only want the broadcasted time, not the (unchanged) `msc_prep`.

**Reply by delahoussou** [reply]
This is the final code: 
```python
msc_prep = xr.open_dataset("msc_prep.nc", engine="netcdf4")["PREC_ACC_NC"].compute()
rainfall_events = msc_prep >= 5
event_labels = np.logical_not(rainfall_events).cumsum(dim="Time").where(rainfall_events) .rename("rainfall_events")
_, time = xr.broadcast(msc_prep, msc_prep.Time)
rain = flox.xarray.xarray_reduce(msc_prep.astype(float), event_labels, func="sum").rename("rainfall_events")
start = flox.xarray.xarray_reduce(time, event_labels, func="min")
end = flox.xarray.xarray_reduce(time, event_labels, func="max")
xr.Dataset({"start": start, "end": end})
```

**Reply by delahoussou** [reply]
It worked but it returned an empty array

**Reply by keewis** [reply]
what does `rainfall_events.sum(dim="Time")` return? If it's all-zero then there's something wrong with the condition / the input data. If not, we will have to debug the algorithm further.

**Reply by delahoussou** [reply]
it is all returning 0

**Reply by delahoussou** [reply]
![A33A048C-F1AF-4E93-8B76-35DCF0D939F8](https://user-images.githubusercontent.com/101612226/228480235-67746b02-a7c8-409e-ba7b-b0155af0d44e.jpeg)

**Reply by keewis** [reply]
okay. How does `msc_prep` look? Does it look plausible? Or are there lots of `nan`s? If so, it's the creation of the cache file (`msc_prep.nc`) that went wrong.

**Reply by delahoussou** [reply]
np.where(msc_prep.values > 0)

(array([], dtype=int64), array([], dtype=int64))

**Reply by delahoussou** [reply]
returning all 0

**Reply by keewis** [reply]
for good measure, could you also just look at `msc_prep` as returned by `open_dataset` (without the `["PREC_ACC_NC"]`) to see if we're pulling out the correct variable?

**Reply by delahoussou** [reply]
Everything is nan, even values of the latittude and longitude

**Reply by delahoussou** [reply]
![3AC77B13-18FD-459B-9E22-7CF4F97BE612](https://user-images.githubusercontent.com/101612226/228482399-284cfe13-9973-4e73-a683-0365ea191daf.jpeg)

**Reply by keewis** [reply]
okay, so that means there definitely is something wrong with the file.

Could you try this:
```python
msc_prep = prsent_prep["PREC_ACC_NC"].isel(
        south_north=xr.DataArray(station_indices[:, 0], dims = "station_indices"),
        west_east=xr.DataArray(station_indices[:, 1], dims = "station_indices"), 
    ).sel(Time=slice("2000", "2013")).compute()
msc_prep
```
to see if that also results in invalid data? If that completes successfully and the data looks fine, you can immediately write it to disk (using `msc_prep.to_netcdf("msc_prep.nc", mode="w", engine="netcdf4")`)

**Reply by delahoussou** [reply]
Should we need to specify engine? else it through an error like in the previous case 
![18D4AC29-DA15-47A2-8773-466A1D69E6F1](https://user-images.githubusercontent.com/101612226/228484348-b6fa99db-1cbf-4081-a50c-4d83b23f15aa.jpeg)

**Reply by keewis** [reply]
can you try running just
```python
msc_prep = prsent_prep["PREC_ACC_NC"].isel(
        south_north=xr.DataArray(station_indices[:, 0], dims = "station_indices"),
        west_east=xr.DataArray(station_indices[:, 1], dims = "station_indices"), 
    ).sel(Time=slice("2000", "2013")).compute()
msc_prep
```
in a cell? That will tell us if this is a read or write error (I suspect it's a read error).

Because that does not involve the writing itself.

**Reply by delahoussou** [reply]
I 
![47DD0D0A-B438-48AC-BF19-152E5E6F581C](https://user-images.githubusercontent.com/101612226/228487092-c229d42a-a439-45ca-a444-da6b23833791.jpeg)
![21E7F227-EBB6-4AFE-B63F-3CB151D82990](https://user-images.githubusercontent.com/101612226/228487095-eb33e7db-829f-4e56-803d-3be54e74a881.jpeg)
![0E3A8C9A-B259-402C-9550-7B8A8505F77F](https://user-images.githubusercontent.com/101612226/228487097-8a602bc2-589e-482e-a9ae-2589eafc00d0.jpeg)
run it in a cell on it own

**Reply by keewis** [reply]
that means your machine is not big enough to handle the chunks we requested. I think I remember that this is also the chunk size on disk, but here's a way to verify that:
```python
# input_file.nc is one of the files read to get `prsent_prep`
ds = xr.open_dataset("input_file.nc", engine="netcdf4")
ds["PREP_ACC_NC"].encoding
```

If that is correct, I'm not sure what to do besides requesting more resources (you only need that while creating `msc_prep.nc`).

**Reply by delahoussou** [reply]
i just copied what you wrote, pasted in another cell and ran it. But it is not working. Talking about memory size I am not running it on my local machine. I am running it on remote server 
jupyterhyb.agron.iastate.edu
![5B9C834D-0036-487E-9CD8-A6ECE44AD065](https://user-images.githubusercontent.com/101612226/228581216-607d8c0d-67c3-4c5f-9f15-9db766a2d0cf.jpeg)

I don't think memory should be an issue there. But this is the error I am getting.

**Reply by keewis** [reply]
sorry, there was a typo: `"netcdf"` → `"netcdf4"`

**Reply by delahoussou** [reply]
should the argument be "input_file.nc" or "msc_prep.nc"
In either case, there is an error
input_filenc -----------------> File not found 
msc_prep.nc ---------------------> KeyError: 'PREP_ACC_NC'

**Reply by keewis** [reply]
`"input_file.nc"` should be one of the files you're reading at the beginning to construct `prsent_prep` (i.e. your original files). Use the actual path for that, `"input_file.nc"` is just a place-holder.

**Reply by delahoussou** [reply]
Ok. So that will be 
fs = '/chinook2/wgallus/climate_output/*_PREC_ACC_NC*.nc'
prsent_prep = xr.open_mfdataset(fs)

**Reply by delahoussou** [reply]
That is what I used.

**Reply by delahoussou** [reply]
I took a single file and I run your code on it. Here is the output
![D50B93A6-372F-4BB6-ABAD-E10486BE9975](https://user-images.githubusercontent.com/101612226/228622773-c2fb558f-cf37-4728-a7f2-1bb98e235c9a.jpeg)

**Reply by keewis** [reply]
great, that means your files are actually chunked (`chunksizes` != `original_shape`), so we can just use a smaller chunksize and everything should be much faster (and the `MemoryError` should go away).

We extract a small number of time series, so we want to have big chunks along time and small ones in space.

To do so, you have to pass `chunks={"south_north": 508, "east_west": 680}` in your `open_mfdataset` call (where you load all the files into `prsent_prep`). I don't know the actual dimension order, so make sure the dimension names match the sizes: the dimension of total size `1015` should have `508` while the dimension of size `1359` should have a chunksize of `680`.

**Reply by delahoussou** [reply]
Ok. Let's recap then 
1. Open the dataset with:
fs = '/chinook2/wgallus/climate_output/*_PREC_ACC_NC*.nc'
prsent_prep = xr.open_mfdataset(fs, chunks={"Time": 1, "south_north": 508, "west_east": 680})
2. Run your code: 
event_labels = np.logical_not(rainfall_events).sum(dim="Time").where(rainfall_events).rename("rainfall_events")
_, time = xr.broadcast(msc_prep, msc_prep.Time)
rain = flox.xarray.xarray_reduce(msc_prep.astype(float), event_labels, func="sum").rename("rainfall_events")
start = flox.xarray.xarray_reduce(time, event_labels, func="min")
end = flox.xarray.xarray_reduce(time, event_labels, func="max")
xr.Dataset({"start": start, "end": end})

**Reply by delahoussou** [reply]
Do I need to delete the msc_prep that we previously created?

**Reply by delahoussou** [reply]
```python
fs = '/chinook2/wgallus/climate_output/*_PREC_ACC_NC*.nc'
prsent_prep = xr.open_mfdataset(fs, chunks={"Time": 1, "south_north": 508, "west_east": 680})["PREC_ACC_NC"].compute()
rainfall_events = msc_prep >= `5`
event_labels = np.logical_not(rainfall_events).sum(dim="Time").where(rainfall_events).rename("rainfall_events")
_, time = xr.broadcast(msc_prep, msc_prep.Time)
rain = flox.xarray.xarray_reduce(msc_prep.astype(float), event_labels, func="sum").rename("rainfall_events")
start = flox.xarray.xarray_reduce(time, event_labels, func="min")
end = flox.xarray.xarray_reduce(time, event_labels, func="max")
xr.Dataset({"start": start, "end": end})
```

**Reply by keewis** [reply]
sorry for the late follow-up. Yes, we don't need the file anymore, now that reading the data should be much quicker. There's a few steps missing to create `msc_prep`, though:
```python
# don't compute `prsent_prep`, that's your entire dataset and most likely won't fit into memory all at once
prsent_prep = xr.open_mfdataset(fs, chunks={"Time": 1, "south_north": 508, "west_east": 680})["PREC_ACC_NC"]
# I think I remember you resampled the time series to every 6 (?) hours. If that's still what you want, do this before the `compute`
# (insert `.resample(Time="6H")` or similar)
msc_prep = (
    prsent_prep.isel(
        south_north=xr.DataArray(station_indices[:, 0], dims = "station_indices"),
        west_east=xr.DataArray(station_indices[:, 1], dims = "station_indices"), 
    )
    .sel(Time=slice("2000", "2013"))
    .compute()
)
rainfall_events = msc_prep >= 5
event_labels = np.logical_not(rainfall_events).sum(dim="Time").where(rainfall_events).rename("rainfall_events")
_, time = xr.broadcast(msc_prep, msc_prep.Time)
rain = flox.xarray.xarray_reduce(msc_prep.astype(float), event_labels, func="sum").rename("rainfall_events")
start = flox.xarray.xarray_reduce(time, event_labels, func="min")
end = flox.xarray.xarray_reduce(time, event_labels, func="max")
result = xr.Dataset({"rain": rain, "start": start, "end": end})
```

**Reply by delahoussou** [reply]
I have was looking at the result. I  am not sure for the date output. It does not seem right. The first output, the start was 2001-02-09 but the end was 2002-06-04. That does not seem right.

**Reply by delahoussou** [reply]
I resample it before saving it. 
![Screenshot (13)](https:
![Screenshot (18)](https://us
![Screenshot (19)](https://user-images.githubusercontent.com/101612226/233730385-ad1f50a2-b019-4d43-b21d-c81abe7e04c8.png)
er-images.githubusercontent.com/101612226/233730381-94d12776-534c-4f97-bee6-b8ff0763e741.png)
//user-images.githubusercontent.com/101612226/233730361-f66a53c9-4845-4467-9396-c19ecfe79e2f.png)
![Screenshot (15)](https://user-images.githubusercontent.com/101612226/233730375-25ba3597-fcc2-49eb-8ba9-09bf1461fbd2.png)

**Reply by keewis** [reply]
so then it works, but not as intended. If that's possible, could you upload the resampled subset (`msc_prep`) as a netcdf file somewhere I can have access to? That would very much speed up the debug cycle.

**Reply by delahoussou** [reply]
Just send me a hi through email, so I have your email. Or if you put your email here I will send send you the access credential. Thanks

**Comment by delahoussou** [comment]
![9E5C7448-2977-4206-AE87-2A9E9E24DE24](https://user-images.githubusercontent.com/101612226/228475456-d114b47a-5bde-4748-8b29-92fd094213e3.jpeg)

**Comment by delahoussou** [comment]
I was looking at the results. I am not sure for the date output. It does not seem right. The first output, the start was 2002-02009 but the corresponding end date was 2002-06-04. That does not seem right. It seems not to be what I was expecting.



________________________________

**Comment by JessicaS11** [comment]
@DOptimusPrime We have a favor to ask. We'd like to point to this discussion thread as an example of [how to utilize xarray's high level computational patterns](https://github.com/xarray-contrib/xarray-tutorial/pull/186). Would you be able to post - as a new comment in this discussion because our link-checker chokes on the GitHub anchor if we try to point to a specific thread post [as in the comment above](https://github.com/pydata/xarray/discussions/7641#discussioncomment-5635306) - the "final" solution that you worked out? Thank you!

**Comment by delahoussou** [comment]
Yes. I can do that.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
